// @framer/framer.default/code/Avatar.tsx
import * as React from "react"
import { addPropertyControls, ControlType, NumberControlDescription, EnumControlDescription } from "framer"
import { defaultEvents, fontControls, fontSizeOptions } from "./lib/constants"

const containerStyle: React.CSSProperties = {
    textAlign: "center",
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center",
    backgroundRepeat: "no-repeat",
    backgroundSize: "cover",
    backgroundPosition: "center center",
}

export function Avatar(props) {
    const {
        onClick,
        onMouseEnter,
        onMouseLeave,
        fontFamily,
        fontWeight,
        fontSize,
        backgroundColor,
        textColor,
        name,
        image,
        width,
    } = props
    const initials = React.useMemo(() => {
        if (!image && name) {
            return getInitials(name)
        }
        return ""
    }, [name, image])

    return (
        <div
            style={{
                ...containerStyle,
                ...{
                    ...(fontFamily && { fontFamily }),
                    fontWeight,
                    fontSize,
                },
                width,
                height: width,
                borderRadius: width,
                backgroundImage: image ? `url(${image})` : undefined,
                backgroundColor: backgroundColor,
                color: textColor,
            }}
            onClick={onClick}
            onMouseEnter={onMouseEnter}
            onMouseLeave={onMouseLeave}
        >
            {initials}
        </div>
    )
}

Avatar.defaultProps = {
    height: 30,
    width: 30,
    name: "Koen Bok",
    textColor: "#fff",
    backgroundColor: "#09F",
    fontWeight: 600,
    fontSize: 11,
}

addPropertyControls(Avatar, {
    name: {
        title: "Name",
        type: ControlType.String,
        defaultValue: Avatar.defaultProps.name,
        placeholder: Avatar.defaultProps.name,
    },
    image: {
        title: "Image",
        type: ControlType.Image,
    },
    textColor: {
        title: "Text Color",
        type: ControlType.Color,
        defaultValue: Avatar.defaultProps.textColor,
    },
    backgroundColor: {
        title: "Bg Color",
        type: ControlType.Color,
        defaultValue: Avatar.defaultProps.backgroundColor,
    },
    ...fontControls,
    fontSize: {
        ...(fontSizeOptions as NumberControlDescription),
        defaultValue: Avatar.defaultProps.fontSize,
    },
    fontWeight: {
        ...(fontControls.fontWeight as EnumControlDescription),
        defaultValue: Avatar.defaultProps.fontWeight,
    },
    ...defaultEvents,
})

function getInitials(name: string, maxInitials = 2) {
    return name
        .split(/\s/)
        .map(part => part.substring(0, 1).toUpperCase())
        .filter(v => !!v)
        .slice(0, maxInitials)
        .join("")
        .toUpperCase()
}
