// @framer/framer.default/code/Time.tsx
import * as React from "react"
import { addPropertyControls, ControlType, NumberControlDescription, EnumControlDescription } from "framer"
import { fontControls, fontSizeOptions, localeOptions, containerStyles } from "./lib/constants"

export function Time(props) {
    const {
        outputType,
        fontFamily,
        fontSize,
        fontWeight,
        localeType,
        customLocale,
        timeFormat,
        showYear,
        showWeekday,
        monthFormat,
    } = props
    const [rerenderOutputKey, setRerenderOutputKey] = React.useState(`${Math.random()}`)
    const timerRef = React.useRef<number>()
    const text = React.useMemo(() => {
        const locale = localeType === "custom" ? [customLocale] : []
        let formatOptions: Intl.DateTimeFormatOptions
        switch (outputType) {
            case "date":
                formatOptions = {
                    weekday: showWeekday ? "long" : undefined,
                    day: "numeric",
                    month: monthFormat,
                    year: showYear ? "numeric" : undefined,
                }
                break
            case "time":
                formatOptions = {
                    hour: "numeric",
                    minute: "numeric",
                    hour12: timeFormat === "12h",
                }
                break
            default:
                console.error(`Unsupported outputType: ${outputType}`)
                break
        }
        return new Intl.DateTimeFormat(locale, formatOptions).format(new Date())
    }, [outputType, localeType, customLocale, timeFormat, showYear, showWeekday, monthFormat, rerenderOutputKey])

    React.useEffect(() => {
        if (outputType === "time") {
            const timer = setInterval(() => setRerenderOutputKey(`${Math.random()}`), 60 - new Date().getSeconds())
            timerRef.current = timer
            return () => {
                if (timer) {
                    return clearInterval(timer)
                }
            }
        }
    }, [timerRef.current, outputType])

    return (
        <div
            style={{
                ...containerStyles,
                ...{
                    ...(fontFamily && { fontFamily }),
                    fontWeight,
                },
                fontSize,
            }}
        >
            {text}
        </div>
    )
}

Time.defaultProps = {
    height: 20,
    width: 140,
    outputType: "time",
    localeType: "auto",
    customLocale: "en-US",
    fontWeight: 500,
    fontSize: 16,
    timeFormat: "24h",
    showYear: true,
    showWeekday: true,
    monthFormat: "long",
}

addPropertyControls(Time, {
    outputType: {
        title: "Type",
        type: ControlType.Enum,
        displaySegmentedControl: true,
        options: ["date", "time"],
        optionTitles: ["Date", "Time"],
        defaultValue: Time.defaultProps.outputType,
    },
    showWeekday: {
        title: "Weekday",
        type: ControlType.Boolean,
        enabledTitle: "Show",
        disabledTitle: "Hide",
        defaultValue: Time.defaultProps.showWeekday,
        hidden: props => props.outputType !== "date",
    },
    monthFormat: {
        title: "Month",
        type: ControlType.Enum,
        options: ["short", "long", "numeric"],
        optionTitles: ["Short", "Long", "Numeric"],
        defaultValue: Time.defaultProps.monthFormat,
        hidden: props => props.outputType !== "date",
    },
    showYear: {
        title: "Year",
        type: ControlType.Boolean,
        enabledTitle: "Show",
        disabledTitle: "Hide",
        defaultValue: Time.defaultProps.showYear,
        hidden: props => props.outputType !== "date",
    },
    timeFormat: {
        title: "Format",
        type: ControlType.Enum,
        options: ["12h", "24h"],
        optionTitles: ["9:14 PM", "21:14"],
        displaySegmentedControl: true,
        defaultValue: Time.defaultProps.timeFormat,
        hidden: props => props.outputType !== "time",
    },
    localeType: {
        title: "Locale",
        type: ControlType.Enum,
        displaySegmentedControl: true,
        options: ["custom", "auto"],
        optionTitles: ["Custom", "Auto"],
        defaultValue: Time.defaultProps.localeType,
    },
    customLocale: {
        title: " ",
        type: ControlType.Enum,
        options: Object.keys(localeOptions).sort(),
        optionTitles: Object.keys(localeOptions)
            .sort()
            .map(key => localeOptions[key]),
        hidden: props => props.localeType !== "custom",
        defaultValue: "en",
    },
    ...fontControls,
    fontSize: {
        ...(fontSizeOptions as NumberControlDescription),
        defaultValue: Time.defaultProps.fontSize,
    },
    fontWeight: {
        ...(fontControls.fontWeight as EnumControlDescription),
        defaultValue: Time.defaultProps.fontWeight,
    },
})
